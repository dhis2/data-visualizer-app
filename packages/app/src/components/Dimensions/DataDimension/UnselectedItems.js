import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { throttle } from 'lodash-es';
import { AssignButton, SelectAllButton } from './buttons';
import { colors } from '../../../colors';

const style = {
    container: {
        border: '1px solid #E0E0E0',
        height: 376,
    },
    listContainer: {
        listStyle: 'none',
        overflowX: 'scroll',
        height: 340,
        width: 418,
        borderBottom: 0,
        paddingLeft: 0,
        margin: 0,
    },
    listItem: {
        display: 'flex',
        margin: 5,
    },
    highlighted: {
        backgroundColor: '#92C9F7',
        borderRadius: 4,
    },
    unHighlighted: {
        display: 'flex',
        padding: 2,
    },
    text: {
        fontFamily: 'Roboto',
        fontSize: 14,
        paddingLeft: 2,
        paddingRight: 2,
    },
    icon: {
        backgroundColor: colors.grey,
        height: 6,
        width: 6,
        marginTop: 4,
        marginLeft: 10,
        marginRight: 5,
    },
};

const UnselectedIcon = () => <div style={style.icon} />;

export class UnselectedItems extends Component {
    constructor(props) {
        super(props);
        this.ulRef = React.createRef();
    }

    state = { highlighted: [] };

    onSelectClick = () => {
        this.props.onSelect(this.state.highlighted);
        this.setState({ highlighted: [] });
    };

    onSelectAllClick = () => {
        this.props.onSelect(this.props.items.map(i => i.id));
        this.setState({ highlighted: [] });
    };

    toggleHighlight = id => {
        const higlightedItems = this.state.highlighted.includes(id)
            ? this.state.highlighted.filter(
                  dataDimId => dataDimId !== id && dataDimId
              )
            : [...this.state.highlighted, id];

        this.setState({ highlighted: higlightedItems });
    };

    onDoubleClickItem = id => this.props.onSelect([id]);

    renderListItem = dataDim => {
        const itemStyle = this.state.highlighted.includes(dataDim.id)
            ? { ...style.unHighlighted, ...style.highlighted }
            : style.unHighlighted;

        return (
            <li
                className="dimension-item"
                key={dataDim.id}
                style={style.listItem}
            >
                <div
                    onDoubleClick={() => this.onDoubleClickItem(dataDim.id)}
                    onClick={() => this.toggleHighlight(dataDim.id)}
                    style={itemStyle}
                >
                    <UnselectedIcon />
                    <span style={style.text}>{dataDim.name}</span>
                </div>
            </li>
        );
    };

    requestMoreItems = throttle(() => {
        const node = this.ulRef.current;

        const bottom = node.scrollHeight - node.scrollTop === node.clientHeight;
        if (bottom) {
            this.props.requestMoreItems();
        }
    }, 1000);

    render = () => {
        const listItems = this.props.items.map(i => this.renderListItem(i));

        return (
            <div style={style.container} onScroll={this.requestMoreItems}>
                <ul ref={this.ulRef} style={style.listContainer}>
                    {listItems}
                </ul>
                <AssignButton action={this.onSelectClick} />
                <SelectAllButton action={this.onSelectAllClick} />
            </div>
        );
    };
}

UnselectedItems.propTypes = {
    items: PropTypes.array.isRequired,
    onSelect: PropTypes.func.isRequired,
    filterText: PropTypes.string.isRequired,
    requestMoreItems: PropTypes.func.isRequired,
};

export default UnselectedItems;
