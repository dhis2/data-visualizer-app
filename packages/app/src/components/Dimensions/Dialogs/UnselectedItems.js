import React, { Component } from 'react';
import PropTypes from 'prop-types';
import throttle from 'lodash-es/throttle';
import { Item } from './Item';
import { AssignButton, SelectAllButton } from './buttons';
import { toggler } from '../../../modules/toggler';

export class UnselectedItems extends Component {
    constructor(props) {
        super(props);
        this.ulRef = React.createRef();
    }

    state = { highlighted: [], lastClickedIndex: 0 };

    onSelectClick = () => {
        this.props.onSelect(this.state.highlighted);
        this.setState({ highlighted: [] });
    };

    onSelectAllClick = () => {
        this.props.onSelect(this.props.items.map(i => i.id));
        this.setState({ highlighted: [] });
    };

    onDoubleClickItem = id => {
        const highlighted = this.state.highlighted.filter(
            dataDimId => dataDimId !== id
        );

        this.setState({ highlighted });
        this.props.onSelect([id]);
    };

    filterTextContains = displayName =>
        displayName.toLowerCase().includes(this.props.filterText.toLowerCase());

    filterItems = (item, index) =>
        this.filterTextContains(item.name)
            ? this.renderListItem(item, index)
            : null;

    toggleHighlight = (isCtrlPressed, isShiftPressed, index, id) => {
        const newState = toggler(
            id,
            isCtrlPressed,
            isShiftPressed,
            index,
            this.state.lastClickedIndex,
            this.state.highlighted,
            this.props.items.map(item => item.id)
        );

        this.setState({
            highlighted: newState.ids,
            lastClickedIndex: newState.lastClickedIndex,
        });
    };

    renderListItem = (dataDim, index) => (
        <li
            className="dimension-item"
            key={dataDim.id}
            onDoubleClick={() => this.onDoubleClickItem(dataDim.id)}
        >
            <Item
                id={dataDim.id}
                index={index}
                displayName={dataDim.name}
                isHighlighted={!!this.state.highlighted.includes(dataDim.id)}
                onItemClick={this.toggleHighlight}
                className={'unselected'}
            />
        </li>
    );

    requestMoreItems = throttle(() => {
        const node = this.ulRef.current;

        if (node) {
            const bottom =
                node.scrollHeight - node.scrollTop === node.clientHeight;
            if (bottom) {
                this.props.requestMoreItems();
            }
        }
    }, 1000);

    render = () => {
        const listItems = this.props.items.map((item, index) =>
            this.props.filterText.length
                ? this.filterItems(item, index)
                : this.renderListItem(item, index)
        );

        return (
            <div
                className={`${this.props.className}-dialog`}
                onScroll={this.requestMoreItems}
            >
                <ul ref={this.ulRef} className={`${this.props.className}-list`}>
                    {listItems}
                </ul>
                <AssignButton
                    className={this.props.className}
                    action={this.onSelectClick}
                />
                <SelectAllButton action={this.onSelectAllClick} />
            </div>
        );
    };
}

UnselectedItems.propTypes = {
    items: PropTypes.array.isRequired,
    onSelect: PropTypes.func.isRequired,
    filterText: PropTypes.string.isRequired,
    requestMoreItems: PropTypes.func,
};

UnselectedItems.defaultProps = {
    requestMoreItems: () => null,
};

export default UnselectedItems;
