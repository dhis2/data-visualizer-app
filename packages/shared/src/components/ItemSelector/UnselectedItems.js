import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import i18n from '@dhis2/d2-i18n';
import throttle from 'lodash-es/throttle';
import Item from './Item';
import { ArrowButton as AssignButton } from './buttons/ArrowButton';
import { SelectButton as SelectAllButton } from './buttons/SelectButton';
import { toggler } from './modules/toggler';
import { styles } from './styles/UnselectedItems.style';

export class UnselectedItems extends Component {
    constructor(props) {
        super(props);
        this.scrolElRef = React.createRef();
    }

    state = { highlighted: [], lastClickedIndex: 0 };

    onSelectClick = () => {
        this.props.onSelect(this.state.highlighted);
        this.setState({ highlighted: [] });
    };

    onSelectAllClick = () => {
        this.props.onSelect(this.props.items.map(i => i.id));
        this.setState({ highlighted: [] });
    };

    onDoubleClickItem = id => {
        const highlighted = this.state.highlighted.filter(
            dataDimId => dataDimId !== id
        );

        this.setState({ highlighted });
        this.props.onSelect([id]);
    };

    filterTextContains = displayName =>
        displayName.toLowerCase().includes(this.props.filterText.toLowerCase());

    filterItems = (item, index) =>
        this.filterTextContains(item.name)
            ? this.renderListItem(item, index)
            : null;

    toggleHighlight = (isCtrlPressed, isShiftPressed, index, id) => {
        const newState = toggler(
            id,
            isCtrlPressed,
            isShiftPressed,
            index,
            this.state.lastClickedIndex,
            this.state.highlighted,
            this.props.items.map(item => item.id)
        );

        this.setState({
            highlighted: newState.ids,
            lastClickedIndex: newState.lastClickedIndex,
        });
    };

    renderListItem = (dataDim, index) => (
        <li
            className="item-selector-item"
            key={dataDim.id}
            onDoubleClick={() => this.onDoubleClickItem(dataDim.id)}
        >
            <Item
                id={dataDim.id}
                index={index}
                name={dataDim.name}
                highlighted={!!this.state.highlighted.includes(dataDim.id)}
                onClick={this.toggleHighlight}
            />
        </li>
    );

    requestMoreItems = throttle(() => {
        const node = this.scrolElRef.current;

        if (node) {
            const bottom =
                node.scrollHeight - node.scrollTop === node.clientHeight;
            if (bottom) {
                this.props.requestMoreItems();
            }
        }
    }, 1000);

    render = () => {
        const listItems = this.props.items.map((item, index) =>
            this.props.filterText.length
                ? this.filterItems(item, index)
                : this.renderListItem(item, index)
        );

        return (
            <Fragment>
                <div
                    ref={this.scrolElRef}
                    onScroll={this.requestMoreItems}
                    style={styles.unselectedItems}
                >
                    <ul className="item-selector-list">{listItems}</ul>
                </div>
                <SelectAllButton
                    style={styles.selectButton}
                    onClick={this.onSelectAllClick}
                    label={i18n.t('Select All')}
                />
                <AssignButton
                    className="item-selector-arrow-forward-button"
                    onClick={this.onSelectClick}
                    iconType={'arrowForward'}
                />
            </Fragment>
        );
    };
}

UnselectedItems.propTypes = {
    items: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            name: PropTypes.string.isRequired,
        })
    ).isRequired,
    onSelect: PropTypes.func.isRequired,
    filterText: PropTypes.string,
    requestMoreItems: PropTypes.func,
};

UnselectedItems.defaultProps = {
    requestMoreItems: () => null,
};

export default UnselectedItems;
