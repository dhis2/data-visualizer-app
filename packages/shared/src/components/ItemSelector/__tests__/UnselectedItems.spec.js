import React from 'react';
import { shallow } from 'enzyme';
import { UnselectedItems } from '../UnselectedItems';

describe('UnselectedItems component', () => {
    let props;
    let shallowItems;
    const unselectedItems = () => {
        if (!shallowItems) {
            shallowItems = shallow(<UnselectedItems {...props} />);
        }
        return shallowItems;
    };

    beforeEach(() => {
        props = {
            items: [],
            filterText: '',
            onSelect: jest.fn(),
            requestMoreItems: jest.fn(),
        };
        shallowItems = undefined;
    });

    it('matches the snapshot when list is empty', () => {
        expect(unselectedItems()).toMatchSnapshot();
    });

    describe('list with items', () => {
        beforeEach(() => {
            props.items = [
                {
                    id: 'rb',
                    name: 'rainbow',
                },
                { id: 'rr', name: 'rarity' },
            ];
        });

        it('matches the snapshot when list has items', () => {
            expect(unselectedItems()).toMatchSnapshot();
        });

        it('triggers onSelect all when "select all" button clicked', () => {
            unselectedItems()
                .find('SelectButton')
                .simulate('click');

            expect(props.onSelect).toHaveBeenCalled();
            expect(props.onSelect).toHaveBeenCalledWith(['rb', 'rr']);
        });

        it('triggers onSelect when item double-clicked', () => {
            unselectedItems()
                .find('li')
                .first()
                .simulate('doubleClick');

            expect(props.onSelect).toHaveBeenCalled();
            expect(props.onSelect).toHaveBeenCalledWith(['rb']);
        });

        it('triggers onSelect when "assign" button clicked', () => {
            const list = unselectedItems();
            list.find('Item')
                .first()
                .simulate('click', false, false, 0, 'rb');

            list.find('ArrowButton')
                .first()
                .simulate('click');

            expect(props.onSelect).toHaveBeenCalled();
            expect(props.onSelect).toHaveBeenCalledWith(['rb']);
        });

        it('renders only items matching the filter', () => {
            props.filterText = 'b';
            const items = unselectedItems().find('Item');

            expect(items.length).toEqual(1);
            expect(items.dive().text()).toEqual(
                expect.stringContaining('rainbow')
            );
        });
    });
});
