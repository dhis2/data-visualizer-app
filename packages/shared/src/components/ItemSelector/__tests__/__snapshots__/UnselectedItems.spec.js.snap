// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnselectedItems component list with items matches the snapshot when list has items 1`] = `
<Fragment>
  <div
    onScroll={[Function]}
    style={
      Object {
        "flex": "1",
        "overflowY": "auto",
      }
    }
  >
    <ul
      className="item-selector-list"
    >
      <li
        className="item-selector-item"
        key="rb"
        onDoubleClick={[Function]}
      >
        <Item
          highlighted={false}
          id="rb"
          index={0}
          name="rainbow"
          onClick={[Function]}
        />
      </li>
      <li
        className="item-selector-item"
        key="rr"
        onDoubleClick={[Function]}
      >
        <Item
          highlighted={false}
          id="rr"
          index={1}
          name="rarity"
          onClick={[Function]}
        />
      </li>
    </ul>
  </div>
  <SelectButton
    label="Select All"
    onClick={[Function]}
    style={
      Object {
        "display": "block",
        "margin": "0 auto",
      }
    }
  />
  <ArrowButton
    className="item-selector-arrow-forward-button"
    iconType="arrowForward"
    onClick={[Function]}
  />
</Fragment>
`;

exports[`UnselectedItems component matches the snapshot when list is empty 1`] = `
<Fragment>
  <div
    onScroll={[Function]}
    style={
      Object {
        "flex": "1",
        "overflowY": "auto",
      }
    }
  >
    <ul
      className="item-selector-list"
    />
  </div>
  <SelectButton
    label="Select All"
    onClick={[Function]}
    style={
      Object {
        "display": "block",
        "margin": "0 auto",
      }
    }
  />
  <ArrowButton
    className="item-selector-arrow-forward-button"
    iconType="arrowForward"
    onClick={[Function]}
  />
</Fragment>
`;
